frontend/expr_inferrer.py

```diff
@@ -0,0 +1,455 @@
+"""Inferrer for python expressions.
+
+Infers the types for the following expressions:
+     - BinOp(expr left, operator op, expr right)
+     - UnaryOp(unaryop op, expr operand)
+     - Dict(expr* keys, expr* values)
+     - Set(expr* elts)
+     - Num(object n)
+     - Str(string s)
+     - NameConstant(singleton value)
+     - List(expr* elts, expr_context ctx)
+     - Tuple(expr* elts, expr_context ctx)
+     - Bytes(bytes s)
+     - IfExp(expr test, expr body, expr orelse)
+	 - Subscript(expr value, slice slice, expr_context ctx)
+	 - Await(expr value) --> Python 3.5+
+	 - Yield(expr? value)
+     - Compare(expr left, cmpop* ops, expr* comparators)
+     - Name(identifier id, expr_context ctx)
+     - FormattedValue(expr value, int? conversion, expr? format_spec) --> Python 3.6+
+     - JoinedStr(expr* values) --> Python 3.6+
+     - ListComp(expr elt, comprehension* generators)
+     - SetComp(expr elt, comprehension* generators)
+     - DictComp(expr key, expr value, comprehension* generators)
+
+     TODO:
+     - Lambda(arguments args, expr body)
+     - GeneratorExp(expr elt, comprehension* generators)
+     - YieldFrom(expr value)
+     - Call(expr func, expr* args, keyword* keywords)
+     - Attribute(expr value, identifier attr, expr_co0ontext ctx)
+     - Starred(expr value, expr_context ctx)
+"""
+
+import ast, sys, predicates as pred
+from i_types import *
+from context import Context
+from abc import ABCMeta, abstractmethod
+from exceptions import HomogeneousTypesConflict
+
+def _filter_types(types, types_filter, *preds):
+    return {t for t in types if (pred.has_supertype(types_filter, t) or pred.satisfies_predicates(t, *preds))}
+
+def narrow_types(original, types_filter, *preds):
```

Comment:
Yes but a small modification of it will be used in the functions inference, so I think it's better to keep it for now.